mathics/core/expression.py

```diff
@@ -901,6 +986,33 @@ def flatten_callback(new_leaves, old):
                 else:
                     return threaded, True
 
+        # Step 6: Now,the next step is to look at the rules associated to
```

Comment:
In https://github.com/Mathics3/mathics-development-guide/blob/3a82527c3461e4db838bd0ded4d86fe304a01df6/docs/extending/code-overview/evaluation.rst#search-for-a-rule-in-head-to-apply I greatly condensed the description of step 6.

A lot of what is being described here is how symbols and parameters get bound. All of this really should be its own method that can be independently tested.